internal/scheduler/scheduler.go

```diff
@@ -0,0 +1,302 @@
+	}
+	return nil
+}
+
+func (srv *Scheduler) updateJobsRuns(ctx context.Context) error {
+	log := ctxlogrus.Extract(ctx)
+	queries := New(srv.Db)
+
+	// New jobs.
+	// TODO: We shouldn't load all columns.
+	// TODO: The first time this runs, we should get only active jobs. After that we should get all jobs.
+	// jobs, err := queries.SelectNewActiveJobs(ctx, srv.JobsSerial)
+	jobs, err := queries.SelectNewJobs(ctx, srv.JobsSerial)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	log.Infof("got updated jobs: %+v", jobs)
+	for _, job := range jobs {
+		job := job
+
+		// Keep the serial number updated.
+		if job.Serial > srv.JobsSerial {
+			srv.JobsSerial = job.Serial
+		}
+
+		// If the job is inactive, remove it from the set of active jobs
+		// and remove all runs associated with the job.
+		if job.Cancelled || job.Succeeded || job.Failed {
+			delete(srv.ActiveJobs, job.JobID)
+			delete(srv.RunsByJobId, job.JobID)
+			continue
+		}
+
+		// If we've never seen this job before, it should be added to the queue.
+		if _, ok := srv.ActiveJobs[job.JobID]; !ok {
+			srv.QueuedJobIds = append(srv.QueuedJobIds, job.JobID)
+		}
+
+		// Add the job to the set of active jobs.
+		srv.ActiveJobs[job.JobID] = &job
+	}
+
+	// New runs for active jobs.
+	i := 0
+	jobIds := make([]uuid.UUID, len(srv.ActiveJobs))
+	for jobId, _ := range srv.ActiveJobs {
+		jobIds[i] = jobId
+		i++
+	}
+	runs, err := queries.SelectNewRunsForJobs(ctx, SelectNewRunsForJobsParams{
+		JobIds: jobIds,
+		Serial: srv.RunsSerial,
+	})
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	for _, run := range runs {
+		run := run
+
+		// Keep the serial number updated.
+		if run.Serial > srv.RunsSerial {
+			srv.RunsSerial = run.Serial
+		}
+
+		jobRuns, ok := srv.RunsByJobId[run.JobID]
+		if !ok {
+			jobRuns = NewJobRuns()
+			srv.RunsByJobId[run.JobID] = jobRuns
+		}
+
+		// Move runs that terminate from the set of active runs to the set of inactive runs,
+		// and consider if the job should be rescheduled.
+		if run.Cancelled || run.Succeeded || run.Failed {
+			delete(jobRuns.ActiveRuns, run.RunID)
+			jobRuns.InactiveRuns[run.RunID] = &run
+			if srv.shouldReScheduleJob(jobRuns) {
+				srv.enqueueJob(run.JobID)
+			}
+			continue
+		}
+	}
+
+	return nil
+}
+
+func (srv *Scheduler) shouldReScheduleJob(jobRuns *JobRuns) bool {
+	// TODO: Look into any previous runs to determine if we should try running the job again.
+	if len(jobRuns.ActiveRuns) == 0 {
+		return false
+	}
+	return false
+}
+
+func (srv *Scheduler) enqueueJob(jobId uuid.UUID) {
+	srv.QueuedJobIds = append(srv.QueuedJobIds, jobId)
+}
+
+func (srv *Scheduler) schedule(ctx context.Context) error {
+	executors := srv.getActiveExecutors()
+	if len(executors) == 0 {
+		if len(srv.Executors) == 0 {
+			return errors.New("no executors available")
+		} else {
+			return errors.Errorf("all executors inactive: %+v", srv.Executors)
+		}
+	}
+
+	// Assign jobs in the queue to executors in a round-robin fashion.
```

Comment:
I presume this is a trivial scheudling algorithm for now (not taking into account priority, capacity etc)?  If so we should put a comment in